File: packages/fundamentals/hooks/native/kotlin/com/pyreon/runtime/PyreonWakeLock.kt
// PyreonWakeLock — the Compose side of `@pyreon/hooks`' useWakeLock.
// Mirror of PyreonWakeLock.swift; see that file's header for why the web arm
// carries normalization this one does not need.

package com.pyreon.runtime

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf

/**
 * The platform half of a wake lock — `FLAG_KEEP_SCREEN_ON` in the real
 * implementation. Swapped for a fake in tests, so the state machine below
 * runs with no Android SDK and no device.
 */
public interface ScreenKeeper {
    /** True when the platform can hold a wake lock at all. */
    public val isSupported: Boolean

    /** Hold or release the screen. Idempotent on both edges. */
    public fun setKeepScreenOn(on: Boolean)
}

/** Reactive screen wake lock — the Compose half of `useWakeLock`. */
public class PyreonWakeLock(private val keeper: ScreenKeeper) {
    public val active: MutableState<Boolean> = mutableStateOf(false)

    public val supported: Boolean get() = keeper.isSupported

    /**
     * Acquire the lock. Returns whether it is now held — false when the
     * platform cannot hold one, matching the web arm's rejected-request path
     * rather than throwing.
     */
    public fun request(): Boolean {
        if (!keeper.isSupported) return false
        if (active.value) return true
        keeper.setKeepScreenOn(true)
        active.value = true
        return true
    }

    /** Release the lock. Safe to call when not held. */
    public fun release() {
        if (!active.value) return
        keeper.setKeepScreenOn(false)
        active.value = false
    }
}

File: packages/fundamentals/hooks/native/swift/PyreonWakeLock.swift
// PyreonWakeLock — the UIKit side of `@pyreon/hooks`' useWakeLock.
//
// ## Why this one is simpler than its web arm
//
// `isIdleTimerDisabled` survives backgrounding: the app resumes with the
// screen still held. The WEB has the awkward half of this contract — a
// WakeLockSentinel is released on tab-hide and never reacquired — so the web
// hook re-acquires on visibilitychange to reach the behaviour this file gets
// for free. The normalization is deliberately one-directional: native is the
// reference, because "the screen stays on until I say otherwise" is what a
// caller means.
//
// ## No UIKit import at this layer
//
// The flag lives behind `IdleTimerController`, so the held/released state
// machine below compiles and runs with no UIKit and no device. `UIKitIdleTimer`
// is the real implementation and is verified on hardware — an approximated
// stub of the display subsystem would prove nothing.

import Foundation
import Observation

/// The platform half of a wake lock. Swapped for a fake in tests.
public protocol IdleTimerController: AnyObject {
    /// True when the platform can hold a wake lock at all.
    var isSupported: Bool { get }
    /// Hold or release the screen. Idempotent on both edges.
    func setIdleTimerDisabled(_ disabled: Bool)
}

/// Observable screen wake lock — the SwiftUI half of `useWakeLock`.
@available(iOS 17.0, macOS 14.0, *)
@Observable
public final class PyreonWakeLock {
    public private(set) var active: Bool = false

    private let controller: IdleTimerController

    public var supported: Bool { controller.isSupported }

    public init(controller: IdleTimerController) {
        self.controller = controller
    }

    /// Acquire the lock. Returns whether it is now held — false when the
    /// platform cannot hold one, matching the web arm's rejected-request
    /// path rather than throwing.
    @discardableResult
    public func request() -> Bool {
        guard controller.isSupported else { return false }
        if active { return true }
        controller.setIdleTimerDisabled(true)
        active = true
        return true
    }

    /// Release the lock. Safe to call when not held.
    public func release() {
        guard active else { return }
        controller.setIdleTimerDisabled(false)
        active = false
    }

    deinit {
        // A lock outliving its view keeps the screen lit with nothing on it.
        if active { controller.setIdleTimerDisabled(false) }
    }
}
